cpp/src/strings/json/json_path.cu

```diff
@@ -0,0 +1,957 @@
+      } break;
+    }
+  }
+
+  return parse_result::SUCCESS;
+}
+
+// hardcoding this for now. to reach a stack depth of 8 would require
+// a jsonpath containing 7 nested wildcards so this is probably reasonable.
+constexpr int max_command_stack_depth = 8;
+
+/**
+ * @brief Parse a single json string using the provided command buffer
+ *
+ * This function exists primarily as a shim for debugging purposes.
+ *
+ * @param input The incoming json string
+ * @param input_len Size of the incoming json string
+ * @param commands The command buffer to be applied to the string. Always ends with a
+ * path_operator_type::END
+ * @param out_buf Buffer user to store the results of the query (nullptr in the size computation
+ * step)
+ * @param out_buf_size Size of the output buffer
+ * @returns A pair containing the result code the output buffer.
+ */
+__device__ thrust::pair<parse_result, json_output> get_json_object_single(
+  char const* input,
+  size_t input_len,
+  path_operator const* const commands,
+  char* out_buf,
+  size_t out_buf_size)
+{
+  json_state j_state(input, input_len);
+  json_output output{out_buf_size, out_buf};
+
+  auto const result = parse_json_path<max_command_stack_depth>(j_state, commands, output);
+
+  return {result, output};
+}
+
+/**
+ * @brief Kernel for running the JSONPath query.
+ *
+ * This kernel operates in a 2-pass way.  On the first pass, it computes
+ * output sizes.  On the second pass it fills in the provided output buffers
+ * (chars and validity)
+ *
+ * @param col Device view of the incoming string
+ * @param commands JSONPath command buffer
+ * @param output_offsets Buffer used to store the string offsets for the results of the query
+ * @param out_buf Buffer used to store the results of the query
+ * @param out_validity Output validity buffer
+ * @param out_valid_count Output count of # of valid bits
+ */
+template <int block_size>
+__launch_bounds__(block_size) __global__
+  void get_json_object_kernel(column_device_view col,
+                              path_operator const* const commands,
+                              size_type* output_offsets,
+                              thrust::optional<char*> out_buf,
+                              thrust::optional<bitmask_type*> out_validity,
+                              thrust::optional<size_type*> out_valid_count)
+{
+  size_type tid    = threadIdx.x + (blockDim.x * blockIdx.x);
+  size_type stride = blockDim.x * gridDim.x;
+
+  if (out_valid_count.has_value()) { *(out_valid_count.value()) = 0; }
+  size_type warp_valid_count{0};
+
+  auto active_threads = __ballot_sync(0xffffffff, tid < col.size());
+  while (tid < col.size()) {
+    bool is_valid         = false;
+    string_view const str = col.element<string_view>(tid);
+    size_type output_size = 0;
+    if (str.size_bytes() > 0) {
+      char* dst = out_buf.has_value() ? out_buf.value() + output_offsets[tid] : nullptr;
```

Comment:
fyi, `optional` is still contextually convertible to bool, so you can still write it like this if you prefer. 
```suggestion
      char* dst = out_buf ? out_buf.value() + output_offsets[tid] : nullptr;
```